docs/tutorials/event_driven.md

```diff
@@ -0,0 +1,298 @@
+  setting the allowedLateness to zero when working with Flink's time windows.
+
+### Performance Considerations
+
+Flink provides `MapState` and `ListState` types that are optimized for RocksDB. Where possible,
+these should be used instead of a `ValueState` object holding some sort of collection. The RocksDB
+state backend can append to `ListState` without going through ser/de, and for `MapState`, each
+key/value pair is a separate RocksDB object, so `MapState` can be efficiently accessed and updated.
+
+{% top %}
+
+## Side Outputs
+
+### Introduction
+
+There are several good reasons to want to have more than one output stream from a Flink operator, such as reporting:
+
+* exceptions
+* malformed events
+* late events
+* operational alerts, such as timed-out connections to external services
+
+Side outputs are a convenient way to do this. 
+
+Each side output channel is associated with an `OutputTag<T>`. The tags have generic types that
+correspond to the type of the side output's DataStream, and they have names. Two OutputTags with the
+same name should have the same type, and will refer to the same side output.
+
+### Example
+
+You are now in a position to do something with the late events that were ignored in the previous
+section. In the `processElement` method of `PseudoWindow` you can now do this:
+
+{% highlight java %}
+if (eventTime <= timerService.currentWatermark()) {
+    // This event is late; its window has already been triggered.
+    OutputTag<TaxiFare> lateFares = new OutputTag<TaxiFare>("lateFares") {};
+    ctx.output(lateFares, fare);
+} else {
+    . . .
+}
+{% endhighlight %}
+
+And the job can access this side output:
+
+{% highlight java %}
+// compute the sum of the tips per hour for each driver
+SingleOutputStreamOperator hourlyTips = fares
+        .keyBy((TaxiFare fare) -> fare.driverId)
+        .process(new PseudoWindow(Time.hours(1)));
+
+OutputTag<TaxiFare> lateFares = new OutputTag<TaxiFare>("lateFares") {};
+hourlyTips.getSideOutput(lateFares).print();
+{% endhighlight %}
+
+{% top %}
+
+## Closing Remarks
+
+In this example you've seen how a ProcessFunction can be used to reimplement a straightforward time
+window. Of course, if Flink's built-in windowing API meets your needs, by all means, go ahead and
+use it. But if you find yourself considering doing something contorted with Flink's windows, don't
+be afraid to roll your own.
+
+Also, ProcessFunctions are useful for many other use cases beyond computing analytics. The hands-on
+exercise below provides an example of something completely different.
+
+Another common use case for ProcessFunctions is for expiring stale state. If you think back to the
+[Rides and Fares Exercise](https://github.com/apache/flink-training/tree/master/rides-and-fares),
+where a `RichCoFlatMapFunction` is used to compute a simple join, the sample solution assumes that
+the TaxiRides and TaxiFares are perfectly matched, one-to-one for each rideId. If an event is lost,
```

Comment:
I'm not sure whether to also put `TaxiFare` and `TaxiRide` into backticks since they append an "s" and that usually renders strangely...
```suggestion
the TaxiRides and TaxiFares are perfectly matched, one-to-one for each `rideId`. If an event is lost,
```